tests/rados/test_pool_osd_recovery.py

```diff
@@ -40,312 +41,471 @@ def run(ceph_cluster, **kw) -> int:
 
-    # Checking cluster health before starting the tests
-    method_should_succeed(rados_obj.run_pool_sanity_check)
-    log.info(
-        "Completed health check of the cluster after test pool creation. Cluster health good!!"
-    )
+            if not ncount_pre < len(
+                rados_obj.run_ceph_command(cmd="ceph orch host ls")
+            ):
+                log.error("New hosts are not added into the cluster")
+                raise Exception("Execution error")
+
+            log.info(
+                "New hosts added to the cluster successfully, Proceeding to deploy OSDs on the same."
+            )
+            # Deploying OSDs on the new nodes.
+            osd_args = {
+                "steps": [
+                    {
+                        "config": {
+                            "command": "apply_spec",
+                            "service": "orch",
+                            "validate-spec-services": True,
+                            "specs": [
+                                {
+                                    "service_type": "osd",
+                                    "service_id": "new_osds",
+                                    "encrypted": "true",
+                                    "placement": {"label": "osd-bak"},
+                                    "spec": {"data_devices": {"all": "true"}},
+                                }
+                            ],
+                        }
+                    }
+                ]
+            }
+            osd_args.update(config)
+            ocount_pre = len(get_osd_details(node=cephadm))
+            add_osd(ceph_cluster=ceph_cluster, config=osd_args)
+            if not ocount_pre < len(get_osd_details(node=cephadm)):
+                log.error("New OSDs were not added into the cluster")
+                raise Exception("Execution error")
+
+            log.info("Deployed new hosts and deployed OSDs on them")
+        except Exception as e:
+            log.error(f"Failed with exception: {e.__doc__}")
+            log.exception(e)
+            raise
+
+    def remove_offline_host(host_node_name: str):
+        # Removing an OSD host and checking status
+        rm_host = get_node_by_id(ceph_cluster, host_node_name)
+        log.info(f"Identified host : {rm_host.hostname} to be removed from the cluster")
+
+        # get list of osd_id on the host to be removed
+        rm_osd_list = rados_obj.collect_osd_daemon_ids(osd_node=rm_host)
+        log.info(f"list of osds on {rm_host.hostname}: {rm_osd_list}")
+        # Starting to drain an offline host.
+        log.info("Starting to drain an offline host")
+        cephadm.shell([f"ceph orch host drain {rm_host.hostname} --force"])
+        # Sleeping for 2 seconds for removal to have started
+        time.sleep(2)
+        log.debug(f"Started drain operation on node : {rm_host.hostname}")
+
+        status_cmd = "ceph orch osd rm status -f json"
```

Comment:
Can we have a re useable method for this? something like `def do_host_drain(host, force=False, wait_till_complete=True, timeout=3600)`

Reply:
The whole test case can be segregated into smaller modules, add_host, remove_host, drain_host
Planning to move all the essential modules to another library workflow
I insist on taking this up as a separate task in another PR

Reply:
Sounds good. Please have a task created for the same.